DEVDOCS-4244: [net new] Storefront GraphQL products and variants (#1625)
bigcommerce/dev-docs: docs/api-docs/storefront/graphql/graphql-variants.md

# This example retrieves 1 variant. 
# Specify multiple variant entityIds to retrieve multiple variants.

query {
  site {
    product (entityId: 113) {           # product entity ID 
      variants (entityIds: [98]) {      # variant entity ID(s)
        edges {
          node {
            # fields on the Variant object type
          }
        }
      }
    }
  }
}
```

<!--
type: tab
title: Variant option value entity ID
-->

```graphql title="Get a variant using variant option values" lineNumbers
# This query retrieves 1 variant.
# You must specify the entity ID for the value of each variant option.

query {
  site {
    product (entityId: 113) {       # product entity ID 
      variants (optionValueIds: [{optionEntityId: 116, valueEntityId: 108} {optionEntityId: 126, valueEntityId: 129}]) {
        edges {
          node {
            # fields on the Variant object type
          }
        }
      }
    }
  }
}
```

&nbsp;

```graphql title="Get a variant using variant option values" lineNumbers
# This query retrieves 2 variants.
# For each variant, you must specify the entity ID for the value of each variant option.

> # This query uses aliases and fragments. For more, see https://graphql.org/learn/queries.

Author: I copyedited a little, though the [text](link) combo doesn't work in code comments

query {
  site {
    product (entityId: 113) {       # product entity ID 
      v1: variants (optionValueIds: [{optionEntityId: 116, valueEntityId: 108} {optionEntityId: 126, valueEntityId: 129}]) {
        edges {
          node {
            ...VariantFields
          }
        }
      }
      v2: variants (optionValueIds: [{optionEntityId: 116, valueEntityId: 109} {optionEntityId: 126, valueEntityId: 129}]){
        edges {
          node {
            ...VariantFields
          }
        }
      }
    }
  }
}

fragment VariantFields on Variant {
  # fields on the Variant object type
}
```

<!-- type: tab-end -->

You can query all variants by not including an argument for `variants`.

### Get a variant with the `product` field

You can also query a variant by specifying a variant identifier for the `product` field. You can use the variant's entity ID, variant option value entity IDs, or variant SKU:

<!--
type: tab
title: Variant entity ID
-->

```graphql title="Get a variant using the variant entity ID" lineNumbers

> # This query retrieves 1 variant. 

Reviewer: ```suggestion
# This query retrieves one variant. 
```

query {
  site {
    product (variantEntityId: 27098) {
      # fields on the Product object type  
    }
  }
}
```

<!--
type: tab
title: Variant option value entity ID
-->

```graphql title="Get a variant using variant option values" lineNumbers

> # This query retrieves 1 variant.

Reviewer: ```suggestion
# This query retrieves one variant.
```

# You must specify the entity ID for the value of each variant option.

query {
  site {
    product (
      entityId: 113       # product entity ID
      optionValueIds: [{optionEntityId: 116, valueEntityId: 108} {optionEntityId: 126, valueEntityId: 129}]
    ) {
      # fields on the Product object type
    }
  }
}

```
<!--
type: tab
title: Variant SKU
-->

```graphql title="Get a variant using the variant sku" lineNumbers

> # This query retrieves 1 variant. 

Reviewer: ```suggestion
# This query retrieves one variant. 
```

query {
  site {
    product (sku: "variant-sku") {
      # fields on the Product object type 
    }
  }
}
```

<!-- type: tab-end -->

Specifying a variant identifier for the `product` field returns variant information overlaid on the Product object. For example, if the variant has a different image, dimensions, SKU, or price, than the product, the variant's info will be returned. This allows you to directly query a variant.

## Get variant identifiers

> You can query identifying info for variants. The following query retrieves identifying info for the specified variant:

Reviewer: ```suggestion
Grab identifying info for variants by using the following query:
```

<!--
type: tab
title: Query
-->

```graphql title="Example query: Get basic info for a variant" lineNumbers
query {
  site {
    product (entityId: 113) {
      variants (entityIds: [127]) {
        edges {
          node {
            id
            entityId
            sku  
            upc
            mpn
          }
        }
      }   
    }
  }
}
```

<!--
type: tab
title: Response
-->

```json title="Example response: Get basic info for a variant" lineNumbers
{
  "data": {
    "site": {
      "product": {
        "variants": {
          "edges": [
            {
              "node": {
                "id": "VmFyaWFudDoxMjc=",
                "entityId": 127,
                "sku": "ABC-1111-PI-SM-CO-RE",
                "upc": "123346",
                "mpn": "9876543210"
              }
            }
          ]
        }
      }
    }
  }
}
```
<!-- type: tab-end -->

## Get variant prices and dimensions

> You can query the prices and dimensions for variants. The following query retrieves prices and dimensions for the specified variant:

Reviewer: ```suggestion
Query the prices and dimensions for variants. The following query retrieves prices and dimensions for the specified variant:
```

<!--
type: tab
title: Query
-->

```graphql title="Example query: Get prices and dimensions for a variant" lineNumbers

> # This query uses fragments. For more, see https://graphql.org/learn/queries/#fragments.

Reviewer: ```suggestion
# This query uses fragments. To learn more about fragments, see GraphQls' documentation segment on [fragments](https://graphql.org/learn/queries).
```

Author: I used the copyedit, though the [text](link) doesn't work in code comments

          "edges": [
            {
              "node": {
                "prices": {
                  "price": {
                    "value": 4.5,
                    "currencyCode": "USD"
                  },
                  "salePrice": {
                    "value": 5,
                    "currencyCode": "USD"
                  },
                  "basePrice": {
                    "value": 10,
                    "currencyCode": "USD"
                  },
                  "retailPrice": {
                    "value": 20,
                    "currencyCode": "USD"
                  }
                },
                "width": {
                  "value": 2,
                  "unit": "in"
                },
                "height": {
                  "value": 2,
                  "unit": "in"
                },
                "depth": {
                  "value": 2,
                  "unit": "in"
                },
                "weight": {
                  "value": 37,
                  "unit": "oz"
                }
              }
            }
          ]
        }
      }
    }
  }
}
```
<!-- type: tab-end -->

## Get variant options and variant option values

> You can query the variant options and the variant option values that are associated with variants. 

Reviewer: ```suggestion
Query the variant options, and their associated values, that are available with variants. 
```

          }
        }
      }
    }
  }
}
```

<!--
type: tab
title: Response
-->

```json title="Example response: Get variant options for a variant" lineNumbers
{
  "data": {
    "site": {
      "product": {
        "variants": {
          "edges": [
            {
              "node": {
                "productOptions": {
                  "edges": [
                    {
                      "node": {
                        "entityId": 116,
                        "displayName": "Color",
                        "isRequired": true,
                        "isVariantOption": true,
                        "displayStyle": "RadioButtons",
                      }
                    },
                    {
                      "node": {
                        "entityId": 126,
                        "displayName": "Size",
                        "isRequired": true,
                        "isVariantOption": true,
                        "displayStyle": "Swatch",
                      }
                    }
                  ]
                }
              }
            }
          ]
        }
      }
    }
  }
}
```
<!-- type: tab-end -->

### Get variant option values

> All variant options are [multiple choice](https://support.bigcommerce.com/s/article/Product-Options-v3?language=en_US#mc). When you get variant options, you can also retrieve the multiple choice values that are associated with a variant (i.e. variant option values). 

Reviewer: ```suggestion
All variant options are [multiple choice](https://support.bigcommerce.com/s/article/Product-Options-v3?language=en_US#mc). With variant options, you get to retrieve the multiple choice values associated with a variant (i.e. variant option values). 
```
